wasm/src/lsp/types.rs

```diff
@@ -0,0 +1,300 @@
+        Url::parse("file:///home/person/stuff.txt").map_err(uniterr).and_then(|uri| {
+            uri.our_to_file_path().map_err(uniterr)
+        }),
+        PathBuf::from_str("/home/person/stuff.txt").map_err(uniterr)
+    );
+}
+
+pub fn cast<R>(req: Request) -> Result<(RequestId, R::Params), ExtractError<Request>>
+where
+    R: lsp_types::request::Request,
+    R::Params: serde::de::DeserializeOwned,
+{
+    req.extract(R::METHOD)
+}
+
+#[derive(Clone, Debug, Hash, PartialEq, Eq, PartialOrd, Ord)]
+pub struct DocPosition {
+    pub line: u32,
+    pub character: u32,
+}
+
+#[derive(Clone, Debug, Hash, PartialEq, Eq, PartialOrd, Ord)]
+pub struct DocRange {
+    pub start: DocPosition,
+    pub end: DocPosition,
+}
+
+#[derive(Clone, Debug)]
+pub struct DocPatch {
+    pub range: DocRange,
+    pub text: String,
+}
+
+impl DocPosition {
+    // This is not dead code, but due to the variation in build configuration
+    // not all ways of building this module require it.
+    // among test, rustlib and wasm32.
+    #[allow(dead_code)]
+    pub fn from_position(pos: &Position) -> Self {
+        DocPosition {
+            line: pos.line,
+            character: pos.character,
+        }
+
+    }
+
+    pub fn to_position(&self) -> Position {
+        Position {
+            line: self.line,
+            character: self.character,
+        }
+    }
+}
+
+impl DocRange {
+    // Not currently used, therefore causing a clippy violation if uncommented.
+    // This can potentially be useful.
+    /*
+    pub fn from_range(r: &Range) -> Self {
+        DocRange {
+            start: DocPosition::from_position(&r.start),
+            end: DocPosition::from_position(&r.end),
+        }
+    }
+    */
+
+    pub fn from_srcloc(l: Srcloc) -> Self {
+        let e = l.ending();
+        DocRange {
+            start: DocPosition {
+                line: if l.line > 0 { (l.line - 1) as u32 } else { 0 },
+                character: if l.col > 0 { (l.col - 1) as u32 } else { 0 },
+            },
+            end: DocPosition {
+                line: if e.line > 0 { (e.line - 1) as u32 } else { 0 },
+                character: if e.col > 0 { (e.col - 1) as u32 } else { 0 },
+            },
+        }
+    }
+
+    pub fn to_srcloc(&self, file: &str) -> Srcloc {
+        let file_rc = Rc::new(file.to_owned());
+        Srcloc::new(
+            file_rc.clone(),
+            (self.start.line + 1) as usize,
+            (self.start.character + 1) as usize,
+        )
+        .ext(&Srcloc::new(
+            file_rc,
+            (self.end.line + 1) as usize,
+            (self.end.character + 1) as usize,
+        ))
+    }
+
+    pub fn to_range(&self) -> Range {
+        Range {
+            start: self.start.to_position(),
+            end: self.end.to_position(),
+        }
+    }
+}
+
+impl DocRange {
+    // This is likewise used in some of the configurations, but not all.
+    #[allow(dead_code)]
+    pub fn overlap(&self, other: &DocRange) -> bool {
```

Comment:
for `overlap`, is there an edge case when creating a DocRange from a Srcloc? I notice that e.g. Srcloc.line == 1 and Srcloc.line == 0 produce the same DocRange, but I'm not sure if it's meaningful to have both of those Srclocs

Reply:
Suggest test near line and col == 0

Reply:
Srcloc is 1 based and DocRange is 0 based.  I guess I'll add tests using invalid Srcloc.